File: examples/ruby/02_validation.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# Validation Example
#
# Demonstrates JSON request body validation, query parameter handling,
# path parameters, and structured error responses.

require 'spikard'
require 'json'

# Create application
app = Spikard::App.new(
  port: 8000,
  host: '127.0.0.1'
)

# Simple in-memory user store
@users = {
  1 => { id: 1, name: 'Alice', email: 'alice@example.com' },
  2 => { id: 2, name: 'Bob', email: 'bob@example.com' }
}

# GET endpoint returning list of users with optional filtering
app.get '/users' do |request|
  users = @users.values

  # Filter by name if query param provided
  if request.query['name']
    name_filter = request.query['name'].downcase
    users = users.select { |u| u[:name].downcase.include?(name_filter) }
  end

  {
    users: users,
    count: users.length
  }
end

# GET endpoint returning single user by ID
app.get '/users/:id' do |request|
  user_id = request.params['id'].to_i

  # Validate ID is a number
  if user_id.zero? && request.params['id'] != '0'
    return {
      status: 400,
      body: {
        error: 'Invalid user ID',
        code: 'invalid_id',
        details: { received: request.params['id'] }
      }
    }
  end

  user = @users[user_id]
  unless user
    return {
      status: 404,
      body: {
        error: 'User not found',
        code: 'not_found',
        details: { user_id: user_id }
      }
    }
  end

  user
end

# POST endpoint to create a new user
app.post '/users' do |request|
  body = request.body

  # Validate required fields
  unless body.is_a?(Hash)
    return {
      status: 400,
      body: {
        error: 'Request body must be JSON',
        code: 'validation_error',
        details: { received_type: body.class.name }
      }
    }
  end

  unless body['name'].is_a?(String) && !body['name'].empty?
    return {
      status: 400,
      body: {
        error: 'Missing or invalid required field: name',
        code: 'validation_error',
        details: { field: 'name' }
      }
    }
  end

  unless body['email'].is_a?(String) && body['email'].include?('@')
    return {
      status: 400,
      body: {
        error: 'Missing or invalid required field: email',
        code: 'validation_error',
        details: { field: 'email' }
      }
    }
  end

  # Create new user
  new_id = @users.keys.max + 1
  new_user = {
    id: new_id,
    name: body['name'],
    email: body['email']
  }
  @users[new_id] = new_user

  {
    status: 201,
    body: new_user
  }
end

# DELETE endpoint to remove user
app.delete '/users/:id' do |request|
  user_id = request.params['id'].to_i

  unless @users[user_id]
    return {
      status: 404,
      body: {
        error: 'User not found',
        code: 'not_found'
      }
    }
  end

  deleted_user = @users.delete(user_id)
  {
    status: 200,
    body: {
      message: 'User deleted',
      user: deleted_user
    }
  }
end

puts 'Starting Validation Example on http://127.0.0.1:8000'
puts 'Try:'
puts '  curl http://127.0.0.1:8000/users'
puts "  curl 'http://127.0.0.1:8000/users?name=Alice'"
puts '  curl http://127.0.0.1:8000/users/1'
puts "  curl -X POST http://127.0.0.1:8000/users \\"
puts "    -H 'Content-Type: application/json' \\"
puts "    -d '{\"name\":\"Charlie\",\"email\":\"charlie@example.com\"}'"
puts ''

app.run

File: examples/node/02-validation.ts
/**
 * Validation Example
 *
 * Demonstrates JSON request body validation, query parameter handling,
 * and structured error responses.
 */

import { get, post, type Request, Spikard } from "@spikard/node";

const app = new Spikard({
	port: 8000,
});

// Simple in-memory user store
const users: Record<number, { id: number; name: string; email: string }> = {
	1: { id: 1, name: "Alice", email: "alice@example.com" },
	2: { id: 2, name: "Bob", email: "bob@example.com" },
};

/**
 * GET endpoint returning a list of users
 */
const listUsers = get("/users")(async function listUsers(req: Request) {
	// Support optional filtering by name via query params
	const nameFilter = req.query?.name as string | undefined;

	let userList = Object.values(users);
	if (nameFilter) {
		userList = userList.filter((u) => u.name.toLowerCase().includes(nameFilter.toLowerCase()));
	}

	return {
		users: userList,
		count: userList.length,
	};
});

/**
 * GET endpoint returning a single user by ID
 */
const getUser = get("/users/:id")(async function getUser(req: Request) {
	const userId = parseInt(req.params?.id as string, 10);

	if (!Number.isInteger(userId)) {
		return {
			status: 400,
			body: {
				error: "Invalid user ID",
				code: "invalid_id",
				details: { received: req.params?.id },
			},
		};
	}

	const user = users[userId];
	if (!user) {
		return {
			status: 404,
			body: {
				error: "User not found",
				code: "not_found",
				details: { user_id: userId },
			},
		};
	}

	return user;
});

/**
 * POST endpoint to create a new user
 */
const createUser = post("/users")(async function createUser(req: Request) {
	const body = req.body as {
		name?: string;
		email?: string;
	} | null;

	// Validate required fields
	if (!body?.name || typeof body.name !== "string") {
		return {
			status: 400,
			body: {
				error: "Missing or invalid required field: name",
				code: "validation_error",
				details: { field: "name", received_type: typeof body?.name },
			},
		};
	}

	if (!body.email || typeof body.email !== "string") {
		return {
			status: 400,
			body: {
				error: "Missing or invalid required field: email",
				code: "validation_error",
				details: { field: "email", received_type: typeof body?.email },
			},
		};
	}

	// Validate email format (simple check)
	if (!body.email.includes("@")) {
		return {
			status: 400,
			body: {
				error: "Invalid email format",
				code: "validation_error",
				details: { field: "email", received: body.email },
			},
		};
	}

	// Create user
	const newId = Math.max(...Object.keys(users).map(Number)) + 1;
	const newUser = {
		id: newId,
		name: body.name,
		email: body.email,
	};
	users[newId] = newUser;

	return {
		status: 201,
		body: newUser,
	};
});

// Register handlers
app.registerHandler(listUsers);
app.registerHandler(getUser);
app.registerHandler(createUser);

console.log("Starting Validation Example on http://127.0.0.1:8000");
console.log("Try:");
console.log("  curl http://127.0.0.1:8000/users");
console.log("  curl 'http://127.0.0.1:8000/users?name=Alice'");
console.log("  curl http://127.0.0.1:8000/users/1");
console.log(
	'  curl -X POST http://127.0.0.1:8000/users -H \'Content-Type: application/json\' -d \'{"name":"Charlie","email":"charlie@example.com"}\'',
);
console.log("");

app.listen().catch((error) => {
	console.error("Server error:", error);
	process.exit(1);
});
